assets/js/components/legacy-notifications/data-error.js

```diff
@@ -65,7 +67,13 @@ function getDataErrorComponent( moduleSlug, errorMessage, inGrid = false, fullWi
 
 	// This is to handle token expired error specifically.
 	if ( 'Invalid Credentials' === errorMessage ) {
-		showErrorNotification( InvalidCredentialsWarning );
+		dispatch( CORE_SITE ).setInternalServerError( {
+			id: 'token-notification',
+			title: __( 'Security Token Error', 'google-site-kit' ),
+			description: __( 'We’re unable to retrieve your data because your security token is expired or revoked.', 'google-site-kit' ),
+			learnMoreURL: global._googlesitekitLegacyData.admin.connectURL,
+			learnMoreLabel: __( 'Please reauthenticate your account', 'google-site-kit' ),
+		} );
```

Comment:
While this is an in-place replacement for what was used before, I don't think this is the right place for such a side-effect. In particular, we want to avoid using the global `Data` registry and always use the provided instance via context.

If we wanted to keep this here, we could instead wrap `cta` with a functional component that used `useDispatch` internally. It should probably also use `useEffectOnce` to avoid dispatching this on every render. We could also use a proper selector for the `connectURL` as well. This would fix the global access of the registry and remove the side-effect in the datastore from calling this function.

Alternatively, we could move this logic into `withData`, particularly into `NewComponent::handleReturnedData`. We would need to also wrap `NewComponent` with `withDispatch` to provide `setInternalServerError` as a prop.

Reply:
@aaemnnosttv I agree with you, this is not perfect, but it's a temporary solution for legacy stuff that will go away soon. Maybe we can leave it as-is for now? I just don't to spend time on rewriting `withData` because it will be a much bigger change that is out of scope for this ticket.

Reply:
Sure, I think the `withData` change would be a bit more out of scope, but what about other idea?

I realize it's temporary but we should avoid reaching for `Data.select` and `Data.dispatch` unless there's really no other way.

<details>
<summary>wrapping CTA with error effect component</summary>

```diff
diff --git a/assets/js/components/legacy-notifications/data-error.js b/assets/js/components/legacy-notifications/data-error.js
index 173255f65..9f63dc94a 100644
--- a/assets/js/components/legacy-notifications/data-error.js
+++ b/assets/js/components/legacy-notifications/data-error.js
@@ -20,19 +20,21 @@
  * WordPress dependencies
  */
 import { __, sprintf } from '@wordpress/i18n';
+import { useEffectOnce } from 'react-use';
 
 /**
  * Internal dependencies
  */
 import Data from 'googlesitekit-data';
 import { CORE_SITE } from '../../googlesitekit/datastore/site/constants';
+import { CORE_USER } from '../../googlesitekit/datastore/user/constants';
 import { getModulesData } from '../../util';
 import { isInsufficientPermissionsError } from '../../util/errors';
 import { getInsufficientPermissionsErrorDescription } from '../../util/insufficient-permissions-error-description';
 import ErrorText from '../ErrorText';
 import CTA from './cta';
 import ctaWrapper from './cta-wrapper';
-const { dispatch } = Data;
+const { useSelect, useDispatch } = Data;
 
 /**
  * Creates a CTA component when there's a data error. Different wrapper HTML is needed depending on where the CTA gets output, which is determined by the inGrid, fullWidth, and createGrid parameters.
@@ -63,17 +65,25 @@ function getDataErrorComponent( moduleSlug, errorMessage, inGrid = false, fullWi
 
 	const reconnectURL = errorObj?.data?.reconnectURL;
 	const description = reconnectURL ? <ErrorText message={ message } reconnectURL={ reconnectURL } /> : message;
-	const cta = <CTA title={ title } description={ description } error />;
+	let cta = <CTA title={ title } description={ description } error />;
 
 	// This is to handle token expired error specifically.
 	if ( 'Invalid Credentials' === errorMessage ) {
-		dispatch( CORE_SITE ).setInternalServerError( {
-			id: 'token-notification',
-			title: __( 'Security Token Error', 'google-site-kit' ),
-			description: __( 'We’re unable to retrieve your data because your security token is expired or revoked.', 'google-site-kit' ),
-			learnMoreURL: global._googlesitekitLegacyData.admin.connectURL,
-			learnMoreLabel: __( 'Please reauthenticate your account', 'google-site-kit' ),
-		} );
+		const CTAWithErrorEffect = () => {
+			const connectURL = useSelect( ( select ) => select( CORE_USER ).getConnectURL() );
+			const { setInternalServerError } = useDispatch( CORE_SITE );
+			useEffectOnce( () => {
+				setInternalServerError( {
+					id: 'token-notification',
+					title: __( 'Security Token Error', 'google-site-kit' ),
+					description: __( 'We’re unable to retrieve your data because your security token is expired or revoked.', 'google-site-kit' ),
+					learnMoreURL: connectURL,
+					learnMoreLabel: __( 'Please reauthenticate your account', 'google-site-kit' ),
+				} );
+			} );
+			return cta;
+		};
+		cta = <CTAWithErrorEffect />;
 	}
 
 	return ctaWrapper( cta, inGrid, fullWidth, createGrid );
```

</details>

Reply:
Ok, I think it can work. Updated, but not sure how to test it :thinking:

Reply:
@eugene-manuilov I just saw that @felixarntz said this could be completely removed on the issue https://github.com/google/site-kit-wp/issues/3406#issuecomment-857116728 🤷‍♂️ 

I've never seen this error before. I'm guessing it's only possible if you were to manually provide an invalid oauth client ID and secret which is more of an edge case for developers or special environments.